src/hotspot/share/prims/jvmtiTagMap.cpp

```diff
@@ -265,8 +249,9 @@ void inline CallbackWrapper::post_callback_tag_update(oop o,
     if (obj_tag == 0) {
       hashmap->remove(o);
     } else {
-      if (obj_tag != entry->tag()) {
-         entry->set_tag(obj_tag);
+      if (obj_tag != current_tag ) {
+        hashmap->remove(o);
+        hashmap->add(o, obj_tag);
```

Comment:
This change is not atomic - is that a problem? The concurrency aspects of using this map are not clear.

Reply:
add() is supposed to add a new object with tag. There is an assert() in its body that checks it. By removing the assert, add() can be used for updating as well.

Reply:
Are you suggesting that `add`  can also act as a `replace` operation? I would think we would want a separate method for that.

Reply:
I didn't notice the suggested name (`replace()`), so It is implemented as `update()` method.